### PR TITLE
explain: further code simplifications & deduplication

### PR DESCRIPTION
**Pull request by aalexandrov**
The next bit of the changes proposed by #24661.

### Motivation

   * This PR refactors existing code.

Mostly removes unneeded code and factors out common functionality in functions / macros.

### Tips for reviewer

* Best review commit by commit, looking at the commit descriptions.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered (Relying on existing tests).
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/adapter/src/optimize/peek.rs [hunk 1]
```diff
@@ -193,22 +193,11 @@ impl Optimize<HirRelationExpr> for Optimizer {
...
 
         // HIR ⇒ MIR lowering and decorrelation
         let expr = expr.lower(&self.config)?;
 
-        // MIR ⇒ MIR optimization (local)
-        self.optimize(expr)
-    }
-}
-
-impl Optimize<MirRelationExpr> for Optimizer {
```
Comment (aalexandrov) [inline src/adapter/src/optimize/peek.rs:208]: This `impl` was needed before I unified the `sequence_` and `explain_` paths in `inner/peek.rs` as part of the off-thread optimization effort. After the second and third commits of #24581 all `peek::Optimizer` instances start with an `HirRelationExpr`.

File: src/adapter/src/optimize/peek.rs [hunk 2]
```diff
@@ -409,7 +398,7 @@ impl<'s> Optimize<GlobalMirPlan<ResolvedGlobal<'s>>> for Optimizer {
             self.config.persist_fast_path_limit,
         )? {
             Some(plan) if self.config.enable_fast_path => {
-                // Trace the pipeline input under `optimize/raw`.
```
Comment (aalexandrov) [inline src/adapter/src/optimize/peek.rs:412]: I committed this with the wrong comment in #24887.

File: src/sql/src/plan.rs [hunk 3]
```diff
@@ -891,27 +891,6 @@ impl ExplaineeStatement {
             Self::CreateIndex { broken, .. } => *broken,
         }
     }
-
-    pub fn row_set_finishing(&self) -> Option<RowSetFinishing> {
```
Comment (aalexandrov) [inline src/sql/src/plan.rs:895]: This is no longer needed after the changes introduced in #24887.

File: src/adapter/src/explain/optimizer_trace.rs [hunk 4]
```diff
@@ -91,6 +94,88 @@ impl OptimizerTrace {
         }
     }
 
+    /// Convert the optimizer trace into a vector or rows that can be returned
+    /// to the client.
+    pub fn into_rows(
```
Comment (aalexandrov) [inline src/adapter/src/explain/optimizer_trace.rs:99]: As suggested in [this comment](https://github.com/MaterializeInc/materialize/pull/24887#discussion_r1476149032).

File: src/adapter/src/optimize/peek.rs [hunk 5]
```diff
@@ -409,7 +407,7 @@ impl<'s> Optimize<GlobalMirPlan<ResolvedGlobal<'s>>> for Optimizer {
...
+                // Trace the pipeline input under `optimize/fast_path`.
                 debug_span!(target: "optimizer", "fast_path").in_scope(|| {
                     trace_plan(&plan);
                 });
```
Comment (teskje) [inline src/adapter/src/optimize/peek.rs:404]: Why is this not using the new macro?

**Review by teskje — APPROVED** [review]
LGTM!

File: src/adapter/src/optimize/peek.rs (hunk 5 above)
Comment (aalexandrov) [reply src/adapter/src/optimize/peek.rs:404]: Ooops.

File: src/adapter/src/optimize/mod.rs [hunk 6]
```diff
@@ -269,3 +269,14 @@ impl From<anyhow::Error> for OptimizerError {
         OptimizerError::Internal(value.to_string())
     }
 }
+
+macro_rules! trace_plan {
+    // match `(at: $span, $plan).
```
[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/adapter/src/optimize/mod.rs]: Removed this.

**Review by ggevay — APPROVED** [review]

**Pull request merged**